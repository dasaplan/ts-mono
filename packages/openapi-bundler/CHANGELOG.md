# @dasaplan/openapi-bundler

## 0.0.7

### Patch Changes

- ef2415e: fixed toposort schemas when array items are allOf objects

## 0.0.6

### Patch Changes

- fix mergeAlOf order to follow default behaviour

## 0.0.5

### Patch Changes

- fix mergeAllOf order to follow default behaviour

## 0.0.3

### Patch Changes

- Update documentation

## 0.0.2

### Patch Changes

- handle dangling properties for schemas with allOf array.
