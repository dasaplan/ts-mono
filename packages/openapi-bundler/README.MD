# @dasaplan/openapi-bundler

- âœ… Configured wrapper of the standard [redocly](https://redocly.com/docs/cli/) bundler.
- âœ… Includes opinionated post-processing steps to transform the spec in a OpenApi 3.0.3 standard compliant way (see [spec-rules][spec-rules]).
- âœ… Is implemented against OpenApi 3.0.3.

## Getting started
### Prerequisite
- using npm
    ```shell
    # integrate in project
    npm i --save-dev @dasaplan/openapi-bundler
    ```    
    ```shell
    # global installation
    npm i -g @dasaplan/openapi-bundler
    ```    
- using pnpm
    ```shell
    # integrate in project
    pnpm i --save-dev @dasaplan/openapi-bundler
    ```    
    ```shell
    # global installation
    pnpm i -g @dasaplan/openapi-bundler
    ```    

### usage
- Bundle an OpenapiSpec
  ```shell
  # assuming the root spec file is located at "$cwd/specs/generic/api.yml"
  # and assuming we want all generated files to find at "$cwd/out"
  openapi-bundler-cli bundle specs/generic/api.yml -o ./out/bundled-api.yml
  ```
- See help for more options
    ```shell
    openapi-bundler-cli bundle --help
    ```
  
## Features
- âœ… Configured wrapper of the standard [redocly](https://redocly.com/docs/cli/) bundler.
> note: Example specs are written in a compressed but valid way to reduce whitespace and facilitate an overview over the whole example.

### post processing
<details>
<summary>âœ… merge allOf subschemas</summary>

<details style="margin-left: 20px"><summary> motivation </summary>
Merging allOf arrays means we reduce complexity for generator and documentation tooling in order to facilitate consistency across tech stacks. 
<p>The easiest way would be to flat out every allOf so there resulting spec does not have the complexity at all. Yet, we need to find a middle ground because codegenerators may rely on identifiers like the discriminator to create object hierarchies.</p> 
We could just let every tool care for themselves how to handle their use-cases. However, from experience this will yield undefined behaviour for API consumer and lead to frustration because the default tooling does no yield results aligning with the specification.

```yaml
# schema declares $ref property but also other properties:
# most tools ignore everything besides $ref
# -> ðŸ”¥ unspecified behavior across tooling and use cases
RefSchema_with_dangling_properties:
  $ref: '#/A_Schema'
  description: "This is something I like to share!"
  title: RefSchema_with_dangling_properties

# schema declares allOf but declares further properties:
# most tool will merge siblings into array, but some may ignore them 
# -> ðŸ”¥ inconsistent behavior across tooling and use cases
AllOfSchema_with_dangling_properties:
  allOf: [ { $ref: '#/A_Schema'} ]
  properties: # sibling to allOf should be part of allOF
    name: { type: string }

# schema declares allOf with multiple elements:
# no problem for validators but code gen mostly try to merge them
# -> ðŸ”¥ merge behaviour may yield unexpected results
AllOfSchema_with_multi_elements:
  allOf:
    - $ref: '#/A_Schema',
    - $ref: '#/B_Schema',
    - properties: { foo: { type: string } }
```
</details>


<details style="margin-left: 20px">
<summary> src spec </summary>

````yaml
components:
  schemas:
    PetBase: 
      type: object,
      properties: { type: { type: string } }
    
    CatBase:  
      type: object,
      discriminator: { propertyName: catType },
      properties: 
        type: { type: string }, 
        catType: { type: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } 

    SeamCat:
      allOf:
        - $ref: '#/components/PetBase'
        - $ref: '#/components/CatBase'
        - title: SeamCat
    ShortHairCat:
      allOf:
        - $ref: '#/components/PetBase'
        - $ref: '#/components/CatBase'
        - title: ShortHairCat

    Dog: { allOf: [ { $ref: '#/components/PetBase' }, { title: Dog } ] }



````
</details>

<details style="margin-left: 20px">
<summary> resulting spec </summary>

````yaml
components:
  schemas:
    PetBase: { type: object, required: [type] discriminator: {propertyName: type},properties: { type: { type: string } } }

    SeamCat:
      allOf:
        - $ref: '#/components/PetBase'
        - { title: SeamCat, type: object, properties: { type: { catType: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } } }

    ShortHairCat:
      allOf:
        - $ref: '#/components/PetBase'
        - { title: ShortHairCat, type: object, properties: { type: { catType: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } } }

    Dog:
      allOf:
        - $ref: '#/components/PetBase'
        - title: Dog 
````
</details>
</details>



<details>
<summary>âœ… ensure discriminator values</summary>

- ensures that:
    - discriminator property type is of type string, and not enum.
        - most sophisticated tools can infer the values from the explicit or implicit discriminator mapping
    - polymorphism and inheritance can be explicitly inferred from the spec
        - every polymorph subschema defines the discriminator property and respective value
            - this is only necessary to extend the standard generator with templating
            - discriminator value is declared with <b>x-const</b> to avoid triggering compatibility layers for discriminator
        - any parent schema referenced from an allOf array does not define discriminator mapping
        - discriminator mapping only exists on schemas with a oneOf member

</details>
<details>
<summary>âœ… support multi schema usage in discriminator mapping (1:n discriminator value to mapping)</summary>

````yaml
Cat:
  type: object
  properties:
    type: 
      type: string
      x-const: [ 'SEAM', 'SHORT_HAIR' ]
    
MyOneOfSchema:
  oneOf: [{$ref: '#/Dog', {$ref: '#/Cat'}}]
  discriminator:
    propertyName: 'type'
    mapping:
      SEAM: "#/Cat"
      SHORT_HAIR: "#/Cat"
      Dog: "#/Dog"
````
</details>
<p></p>
<details>
<summary> <b style="color: #2c6fdf">ðŸ“œ complete example: openapi post-processing </b> </summary>

| examples | src                                 | bundled                                     | post-processed                                       |
|----------|-------------------------------------|---------------------------------------------|------------------------------------------------------|
| simple   | [link to src][simple-petstore-src]  | [link to bundled][simple-petstore-bundled]  | [link to post-processed][simple-petstore-processed]  |
| complex  | [link to src][complex-petstore-src] | [link to bundled][complex-petstore-bundled] | [link to post-processed][complex-petstore-processed] |

</details>

[complex-petstore-src]: ../openapi-specs/specs/pets-modular-complex/petstore-api.yml

[complex-petstore-bundled]: ./docs/examples/specs/complex-petstore-bundled.yml

[complex-petstore-processed]: ./docs/examples/specs/complex-petstore-post-processed.yml

[simple-petstore-src]: ../openapi-specs/specs/pets-simple/pets-api.yml

[simple-petstore-bundled]: ./docs/examples/specs/simple-petstore-bundled.yml

[simple-petstore-processed]: ./docs/examples/specs/simple-petstore-post-processed.yml

[spec-rules]: ./docs/spec-rules.md