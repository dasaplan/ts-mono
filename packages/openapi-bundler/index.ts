export {
  bundleOpenapi,
  createCommandBundle,
  createSpecProcessor,
  cleanObj,
  OpenApiBundled,
  parseOpenapi,
  Resolver,
  Transpiler,
  OaComponent,
  SchemaResolverContext,
  Schema,
} from "./src/index.js";
