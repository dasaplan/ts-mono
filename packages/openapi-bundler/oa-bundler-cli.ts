#!/usr/bin/env ts-node

import { Command } from "commander";
import * as process from "process";
import { createCommandBundle } from "./src/index.js";

const program = new Command();
createCommandBundle(program);

program.parse(process.argv);
