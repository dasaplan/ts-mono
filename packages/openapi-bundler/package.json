{
  "name": "@dasaplan/openapi-bundler",
  "description": "Opinionated bundler/merger for Openapi (Swagger)",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "version": "0.0.0",
  "keywords": [
    "openapi",
    "typescript",
    "ts",
    "zod",
    "codegen"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/dasaplan/ts-mono.git",
    "directory": "packages/openapi-bundler"
  },
  "scripts": {
    "prepack": "sort-package-json --quiet && clean-package",
    "postpack": "clean-package restore",
    "build": "pnpm rup",
    "rup": "tsup-node --dts",
    "lint": "eslint \"./src/*.ts\" index.ts oa-bundler-cli.ts",
    "clean": "rimraf dist",
    "test": "pnpm test:esm",
    "test:esm": "cross-env NODE_OPTIONS=--experimental-vm-modules jest",
    "test:esm:update": "cross-env NODE_OPTIONS=--experimental-vm-modules jest -u"
  },
  "bin": {
    "openapi-bundler-cli": "dist/oa-cli.js"
  },
  "files": ["dist"],
  "dependencies": {
    "commander": "^11.1.0",
    "json-schema-merge-allof": "^0.8.1",
    "jsonpointer": "^5.0.1",
    "openapi3-ts": "^4.3.3",
    "merge-json-schemas": "^1.0.0",
    "tslog": "^4.9.2",
    "lodash": "^4.17.21",
    "toposort": "^2.0.2"
  },
  "peerDependencies": {
    "@redocly/openapi-core": "^1.6.0"
  },
  "devDependencies": {
    "tsconfig": "workspace:*",
    "@types/json-schema-merge-allof": "^0.6.5",
    "@types/lodash":"^4.17.5",
    "@dasaplan/ts-sdk": "workspace:*"
  },
  "clean-package": {
    "remove": [
      "scripts",
      "devDependencies",
      "tsup",
      "clean-package",
      "dependencies.@dasaplan/ts-sdk"
    ]
  },
  "tsup": {
    "entry": [
      "index.ts",
      "oa-bundler-cli.ts"
    ],
    "noExternal": [
      "@dasaplan/ts-sdk",
      "tsconfig"
    ],
    "format": [
      "cjs",
      "esm"
    ],
    "tsconfig": "tsconfig.build.json",
    "clean": true,
    "sourcemap": true
  }
}
