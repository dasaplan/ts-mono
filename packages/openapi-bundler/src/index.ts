export { bundleOpenapi, OpenApiBundled, parseOpenapi } from "./bundle.js";
export { createCommandBundle } from "./commands.js";
export { createSpecProcessor } from "./post-process/index.js";
export { Transpiler, Schema } from "./transpiler/index.js";
export {
  Resolver,
  OaComponent,
  SchemaResolverContext,
  cleanObj,
} from "./resolver/index.js";
