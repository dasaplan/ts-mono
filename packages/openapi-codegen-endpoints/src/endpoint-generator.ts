/* eslint-disable @typescript-eslint/no-unused-vars */
import { appLog } from "./logger.js";
import { bundleParseOpenapi, OpenApiBundled } from "@dasaplan/openapi-bundler";
import { _, Folder, File } from "@dasaplan/ts-sdk";
import { Project } from "ts-morph";
import { Templates } from "./templates.js";
import { EndpointInterfaceGeneratorOptions, generateEndpointInterfacesAsText } from "./endpoint-interfaces.js";
import { pascalCase } from "pascal-case";
import { createTsMorphSrcFile, createTsMorphSrcFileFromText } from "./ts-sources.js";

export interface EndpointDefinitionGeneratorOptions extends EndpointInterfaceGeneratorOptions {
  outDir: string;
  templatesDir?: string;
}

export async function generateEndpointDefinitions(openapiSpec: string, params: EndpointDefinitionGeneratorOptions) {
  appLog.childLog(generateEndpointDefinitions).info(`start generate:`, openapiSpec);
  const bundled = await bundleParseOpenapi(openapiSpec, { mergeAllOf: true, ensureDiscriminatorValues: true });
  return await generateEndpointDefinitionsFromBundled(bundled, params);
}

export async function generateEndpointDefinitionsFromBundled(bundled: OpenApiBundled, params: EndpointDefinitionGeneratorOptions) {
  const apiName = createApiName(bundled, params);
  const endpoints = await generateEndpointInterfacesAsText(bundled, { ...params, apiName });
  const withImports = ["import {EndpointDefinition} from './EndpointDefinition'", endpoints].join("\n");

  const out = Folder.of(params?.outDir ?? "out").create();
  const { project } = await generateTemplates({ ...params, outDir: out.absolutePath });
  const endpointFileName = `${apiName}.ts`;
  const endpointFilePath = out.makeFile(endpointFileName).absolutePath;
  createTsMorphSrcFileFromText(endpointFilePath, withImports, project);
  await project.save();

  return { endpointFileName, endpointFilePath, sources: project.getSourceFiles().map((s) => s.getText()) };
}

export function createApiName(bundled: OpenApiBundled, params: EndpointDefinitionGeneratorOptions) {
  const rawApiName = params.apiName ?? bundled.info.title;
  const apiName = pascalCase(rawApiName);
  const name = _.isEmpty(apiName) ? "Api" : apiName;
  return `${name}Endpoints`;
}

async function generateTemplates(params: EndpointDefinitionGeneratorOptions, project: Project = new Project()) {
  const endpointTmpl = Templates.getTemplateFile("EndpointDefinition.d.ts");
  const source = createTsMorphSrcFile(endpointTmpl.absolutePath, project);

  const outFile = File.resolve(params.outDir, endpointTmpl.name);
  source.sourceFile.copy(outFile.absolutePath, { overwrite: true });
  return { project, sourceFile: source.sourceFile };
}
