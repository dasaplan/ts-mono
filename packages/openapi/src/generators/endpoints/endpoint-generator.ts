/* eslint-disable @typescript-eslint/no-unused-vars */
import { appLog } from "../../logger.js";
import { bundleParseOpenapi, Endpoint, OpenApiBundled, Transpiler } from "@dasaplan/openapi-bundler";
import { Project } from "ts-morph";
import path from "node:path";

export interface EndpointDefinitionOptions {}

// export function generateEndpointDefinitions(openapiSpec: string, out: string, _params?: EndpointDefinitionOptions) {
//   appLog.childLog(generateEndpointDefinitions).info(`start generate:`, openapiSpec, out);
//
//   const outDir = path.isAbsolute(out) ? out : path.resolve(process.cwd(), out);
//
//   const _apiFile = File.of(outDir, "endpoints.ts");
//
//   return outDir;
// }

function createEndpointDefinition(endpoint: Endpoint) {
  return `
    
    `;
}

export async function generateEndpointDefinitionsInMemory(openapiSpec: string, _params?: EndpointDefinitionOptions) {
  appLog.childLog(generateEndpointDefinitionsInMemory).info(`start generate:`, openapiSpec);
  const bundled = await bundleParseOpenapi(openapiSpec, { mergeAllOf: true, ensureDiscriminatorValues: true });
  return generateEndpointDefinitions(bundled);
}
export async function generateEndpointDefinitions(bundled: OpenApiBundled, _params?: EndpointDefinitionOptions) {
  const endpoints = Transpiler.of(bundled).endpoints();
  const common = JSON.stringify();
  return endpoints.map(createEndpointDefinition);
}

function createTsMorphSrcFile(tsFilePath: string) {
  const project = new Project();
  project.createSourceFile(tsFilePath, "");
  const sourceFile = project.getSourceFile(path.basename(tsFilePath));
  if (_.isNil(sourceFile)) {
    throw `Error: Expected source file for provided path: srcFile: ${tsFilePath}`;
  }
  return { project, sourceFile: sourceFile };
}
