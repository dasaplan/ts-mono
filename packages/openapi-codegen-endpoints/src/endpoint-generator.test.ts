/* eslint-disable @typescript-eslint/no-unused-vars */
import { OpenApiBundled } from "@dasaplan/openapi-bundler";
import { oas30 } from "openapi3-ts";
import { generateEndpointDefinitions } from "./endpoint-generator.js";
import { describe, expect, test } from "vitest";

describe("generateEndpointDefinitions", () => {
  test("circular schema", async () => {
    const openapi: OpenApiBundled = createApi(
      (oa) =>
        withSchemas(oa, {
          ResponseSchema: {
            type: "object",
            properties: {
              id: { type: "string" },
            },
          },
        }),
      (oa) =>
        withRoute(oa, {
          "/pets/{:petId}": {
            get: {
              operationId: "getPet",
              parameters: [{ in: "path", name: "petId", required: true, schema: { type: "string" } }],
              responses: {
                200: {
                  content: { "application/json": { schema: { $ref: "#/components/schemas/ResponseSchema" } } },
                },
                201: {
                  description: "success",
                },
                401: {
                  content: { "application/json": { schema: { type: "object", properties: { message: { type: "string" } } } } },
                },
              },
            },
            put: {
              operationId: "updatePet",
              parameters: [{ in: "query", name: "secret", required: true, schema: { type: "string" } }],
              requestBody: {
                content: {
                  "application/json": {
                    schema: { type: "object", properties: { help: { type: "string" } } },
                  },
                },
              },
              responses: {
                200: {
                  content: "application/json",
                  schema: { $ref: "#/components/schemas/ResponseSchema" },
                },
              },
            },
          },
          "/pets": {
            post: {
              operationId: "createPet",
              parameters: [
                { in: "cookie", name: "secret", required: true, schema: { type: "string" } },
                { in: "header", name: "other-secret", required: true, schema: { type: "string" } },
              ],
              requestBody: {
                content: {
                  "application/json": {
                    schema: { type: "object", properties: { help: { type: "string" } } },
                  },
                },
              },
              responses: {
                200: {
                  content: "application/json",
                  schema: { $ref: "#/components/schemas/ResponseSchema" },
                },
              },
            },
          },
        })
    );

    const endpoints = await generateEndpointDefinitions(openapi);

    expect(endpoints).toMatchInlineSnapshot(`
      [
        "
          
          ",
        "
          
          ",
        "
          
          ",
      ]
    `);
  });
});

function createApi(...mods: Array<(oa: OpenApiBundled) => OpenApiBundled>): OpenApiBundled {
  const api: OpenApiBundled = {
    openapi: "3.0.3",
    info: { version: "", title: "" },
    paths: {},
    components: {
      schemas: {},
    },
  };
  return mods.reduce((acc, curr) => curr(acc), api);
}

function withSchemas(oa: OpenApiBundled, schemas: NonNullable<oas30.ComponentsObject["schemas"]>): OpenApiBundled {
  return {
    ...oa,
    components: {
      ...oa.components,
      schemas: {
        ...(oa.components?.schemas ?? {}),
        ...schemas,
      },
    },
  };
}
function withRoute(oa: OpenApiBundled, routes: NonNullable<oas30.PathsObject>): OpenApiBundled {
  return {
    ...oa,
    paths: {
      ...oa.paths,
      ...routes,
    },
  };
}
