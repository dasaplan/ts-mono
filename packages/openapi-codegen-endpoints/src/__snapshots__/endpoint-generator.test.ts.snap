// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateEndpointDefinitions > enspoints 1`] = `
[
  "export interface EndpointDefinition<
    DeserializedResponse extends EndpointDefinition.DeserializedResponsesObject,
    DeserializedRequest extends EndpointDefinition.DeserializedRequestObject,
    Params extends EndpointDefinition.Parameters | undefined
> {
    path: EndpointDefinition.Path;
    /** operation id */
    name: string;
    operation: EndpointDefinition.Operation;
    parameters?: Params;
    response: EndpointDefinition.Response<DeserializedResponse>;
    request?: EndpointDefinition.Request<DeserializedRequest>;
}

export module EndpointDefinition {
    type PathSegment = \`\${string}\`;

    type PathParam = \`{:\${string}}\`;

    /** @example '/pets', '/pets/{:petId}' , '/pets/{:petId}/inner-resources' */
    export type Path = \`/\${PathSegment}\` | \`/\${PathSegment}/\${PathSegment | PathParam}\`;

    type DtoTypes = object | string | number | undefined;

    export type DeserializedResponsesObject = { [status: number]: DtoTypes };

    export type Operation = "get" | "post" | "put" | "delete" | "patch";

    export type Parameters = {
        path?: { [name: string]: DtoTypes };
        query?: { [name: string]: DtoTypes };
        header?: { [headerKey: string]: DtoTypes };
        cookie?: { [cookieKey: string]: DtoTypes };
    };
    export type DeserializedRequestObject = DtoTypes;

    export interface Request<DeserializedRequest = DeserializedRequestObject> {
        /** meta information */
        format?: string;
        /** deserialize payload */
        payload: DeserializedRequest;
        /** transform dto to model object */
        transform?: <Model, Dto extends DeserializedRequest>(payload: Dto) => Model | undefined;
        deserialize?: (payload: string | number | undefined) => DeserializedRequest;
    }

    export type Response<DeserializedResponse = DeserializedResponsesObject> = {
        [status in keyof DeserializedResponse]: {
            /** meta information */
            format?: string;
            /** deserialize payload */
            payload: DeserializedResponse[status];
            /** transform model object to dto */
            transform?: <Model>(model: Model) => DeserializedResponse[status];
        };
    };
}
",
  "export interface EndpointDefinition<
    DeserializedResponse extends EndpointDefinition.DeserializedResponsesObject,
    DeserializedRequest extends EndpointDefinition.DeserializedRequestObject,
    Params extends EndpointDefinition.Parameters | undefined
> {
    path: EndpointDefinition.Path;
    /** operation id */
    name: string;
    operation: EndpointDefinition.Operation;
    parameters?: Params;
    response: EndpointDefinition.Response<DeserializedResponse>;
    request?: EndpointDefinition.Request<DeserializedRequest>;
}

export module EndpointDefinition {
    type PathSegment = \`\${string}\`;

    type PathParam = \`{:\${string}}\`;

    /** @example '/pets', '/pets/{:petId}' , '/pets/{:petId}/inner-resources' */
    export type Path = \`/\${PathSegment}\` | \`/\${PathSegment}/\${PathSegment | PathParam}\`;

    type DtoTypes = object | string | number | undefined;

    export type DeserializedResponsesObject = { [status: number]: DtoTypes };

    export type Operation = "get" | "post" | "put" | "delete" | "patch";

    export type Parameters = {
        path?: { [name: string]: DtoTypes };
        query?: { [name: string]: DtoTypes };
        header?: { [headerKey: string]: DtoTypes };
        cookie?: { [cookieKey: string]: DtoTypes };
    };
    export type DeserializedRequestObject = DtoTypes;

    export interface Request<DeserializedRequest = DeserializedRequestObject> {
        /** meta information */
        format?: string;
        /** deserialize payload */
        payload: DeserializedRequest;
        /** transform dto to model object */
        transform?: <Model, Dto extends DeserializedRequest>(payload: Dto) => Model | undefined;
        deserialize?: (payload: string | number | undefined) => DeserializedRequest;
    }

    export type Response<DeserializedResponse = DeserializedResponsesObject> = {
        [status in keyof DeserializedResponse]: {
            /** meta information */
            format?: string;
            /** deserialize payload */
            payload: DeserializedResponse[status];
            /** transform model object to dto */
            transform?: <Model>(model: Model) => DeserializedResponse[status];
        };
    };
}
",
  "import { EndpointDefinition } from './EndpointDefinition'
export module TestApiEndpoints {
    export type Path = "/pets/{:petId}" | "/pets"
    export interface OperationToPath {
        getPet: "/pets/{:petId}";
        updatePet: "/pets/{:petId}";
        createPet: "/pets";
    }
    export interface GetPet<ErrorSchema extends EndpointDefinition.DtoTypes> extends EndpointDefinition<
        { "401": ErrorSchema },
        undefined,
        { "path": { "petId": string }, "query": undefined, "header": undefined, "cookie": undefined }
    > {
        name: "getPet";
        operation: "get";
        path: "/pets/{:petId}"
    }

    export interface UpdatePet<ResponseSchema extends EndpointDefinition.DtoTypes, RequestSchema extends EndpointDefinition.DtoTypes> extends EndpointDefinition<
        { "200": ResponseSchema },
        RequestSchema,
        { "path": undefined, "query": { "secret": string }, "header": undefined, "cookie": undefined }
    > {
        name: "updatePet";
        operation: "put";
        path: "/pets/{:petId}"
    }

    export interface CreatePet<RequestSchema extends EndpointDefinition.DtoTypes> extends EndpointDefinition<
        { "200": undefined },
        RequestSchema,
        { "path": undefined, "query": undefined, "header": { "other-secret": string }, "cookie": { "secret": string } }
    > {
        name: "createPet";
        operation: "post";
        path: "/pets"
    }

}
",
]
`;

exports[`generateEndpointDefinitions > integration > pets-modular-complex/petstore-api.yml 1`] = `
[
  "export interface EndpointDefinition<
    DeserializedResponse extends EndpointDefinition.DeserializedResponsesObject,
    DeserializedRequest extends EndpointDefinition.DeserializedRequestObject,
    Params extends EndpointDefinition.Parameters | undefined
> {
    path: EndpointDefinition.Path;
    /** operation id */
    name: string;
    operation: EndpointDefinition.Operation;
    parameters?: Params;
    response: EndpointDefinition.Response<DeserializedResponse>;
    request?: EndpointDefinition.Request<DeserializedRequest>;
}

export module EndpointDefinition {
    type PathSegment = \`\${string}\`;

    type PathParam = \`{:\${string}}\`;

    /** @example '/pets', '/pets/{:petId}' , '/pets/{:petId}/inner-resources' */
    export type Path = \`/\${PathSegment}\` | \`/\${PathSegment}/\${PathSegment | PathParam}\`;

    type DtoTypes = object | string | number | undefined;

    export type DeserializedResponsesObject = { [status: number]: DtoTypes };

    export type Operation = "get" | "post" | "put" | "delete" | "patch";

    export type Parameters = {
        path?: { [name: string]: DtoTypes };
        query?: { [name: string]: DtoTypes };
        header?: { [headerKey: string]: DtoTypes };
        cookie?: { [cookieKey: string]: DtoTypes };
    };
    export type DeserializedRequestObject = DtoTypes;

    export interface Request<DeserializedRequest = DeserializedRequestObject> {
        /** meta information */
        format?: string;
        /** deserialize payload */
        payload: DeserializedRequest;
        /** transform dto to model object */
        transform?: <Model, Dto extends DeserializedRequest>(payload: Dto) => Model | undefined;
        deserialize?: (payload: string | number | undefined) => DeserializedRequest;
    }

    export type Response<DeserializedResponse = DeserializedResponsesObject> = {
        [status in keyof DeserializedResponse]: {
            /** meta information */
            format?: string;
            /** deserialize payload */
            payload: DeserializedResponse[status];
            /** transform model object to dto */
            transform?: <Model>(model: Model) => DeserializedResponse[status];
        };
    };
}
",
  "export interface EndpointDefinition<
    DeserializedResponse extends EndpointDefinition.DeserializedResponsesObject,
    DeserializedRequest extends EndpointDefinition.DeserializedRequestObject,
    Params extends EndpointDefinition.Parameters | undefined
> {
    path: EndpointDefinition.Path;
    /** operation id */
    name: string;
    operation: EndpointDefinition.Operation;
    parameters?: Params;
    response: EndpointDefinition.Response<DeserializedResponse>;
    request?: EndpointDefinition.Request<DeserializedRequest>;
}

export module EndpointDefinition {
    type PathSegment = \`\${string}\`;

    type PathParam = \`{:\${string}}\`;

    /** @example '/pets', '/pets/{:petId}' , '/pets/{:petId}/inner-resources' */
    export type Path = \`/\${PathSegment}\` | \`/\${PathSegment}/\${PathSegment | PathParam}\`;

    type DtoTypes = object | string | number | undefined;

    export type DeserializedResponsesObject = { [status: number]: DtoTypes };

    export type Operation = "get" | "post" | "put" | "delete" | "patch";

    export type Parameters = {
        path?: { [name: string]: DtoTypes };
        query?: { [name: string]: DtoTypes };
        header?: { [headerKey: string]: DtoTypes };
        cookie?: { [cookieKey: string]: DtoTypes };
    };
    export type DeserializedRequestObject = DtoTypes;

    export interface Request<DeserializedRequest = DeserializedRequestObject> {
        /** meta information */
        format?: string;
        /** deserialize payload */
        payload: DeserializedRequest;
        /** transform dto to model object */
        transform?: <Model, Dto extends DeserializedRequest>(payload: Dto) => Model | undefined;
        deserialize?: (payload: string | number | undefined) => DeserializedRequest;
    }

    export type Response<DeserializedResponse = DeserializedResponsesObject> = {
        [status in keyof DeserializedResponse]: {
            /** meta information */
            format?: string;
            /** deserialize payload */
            payload: DeserializedResponse[status];
            /** transform model object to dto */
            transform?: <Model>(model: Model) => DeserializedResponse[status];
        };
    };
}
",
  "import { EndpointDefinition } from './EndpointDefinition'
export module TestApiEndpoints {
    export type Path = "/pets" | "/pets/{petId}"
    export interface OperationToPath {
        listPets: "/pets";
        createPets: "/pets";
        showPetById: "/pets/{petId}";
        patchPet: "/pets/{petId}";
    }
    export interface ListPets<Array<Pet> extends EndpointDefinition.DtoTypes > extends EndpointDefinition <
        { "200": Array<Pet> },
        undefined,
        { "path": { "petId": string }, "query": undefined, "header": undefined, "cookie": undefined }
        > {
            name: "listPets";
            operation: "get";
            path: "/pets"
        }

    export interface CreatePets<Pet extends EndpointDefinition.DtoTypes> extends EndpointDefinition<
        {},
        Pet,
        { "path": undefined, "query": undefined, "header": undefined, "cookie": undefined }
    > {
        name: "createPets";
        operation: "post";
        path: "/pets"
    }

    export interface ShowPetById<Pet extends EndpointDefinition.DtoTypes> extends EndpointDefinition<
        { "200": Pet },
        undefined,
        { "path": { "petId": string }, "query": undefined, "header": undefined, "cookie": undefined }
    > {
        name: "showPetById";
        operation: "get";
        path: "/pets/{petId}"
    }

    export interface PatchPet<Pet extends EndpointDefinition.DtoTypes, Pet extends EndpointDefinition.DtoTypes> extends EndpointDefinition<
        { "200": Pet },
        Pet,
        { "path": { "petId": string }, "query": undefined, "header": undefined, "cookie": undefined }
    > {
        name: "patchPet";
        operation: "patch";
        path: "/pets/{petId}"
    }

}
",
]
`;
